File: packages/react-native/ios/HttpClient.swift
import Foundation
import Alamofire

class HttpClient {
    static let shared = HttpClient()
    private var baseUri: String?
    private var publicKey: String?
    private var headers: Dictionary<String, String> = [:]
    private var cacheTimeout: Int?
    private var requestQueueLimit: Int = 5
    private var requestQueue: Queue<DashXRequest<T>>
    
    private init(_ baseUri: String? = nil, _ publicKey: String? = nil) {
        self.baseUri = baseUri
        self.publicKey = publicKey
    }
    
    func create() -> HttpClient {
        return HttpClient(self.baseUri, self.publicKey)
    }
    
    func withBaseUri(_ baseUri: String) -> HttpClient {
        self.baseUri = baseUri
        return self
    }
    
    func withPublicKey(_ publicKey: String) -> HttpClient {
        self.publicKey = publicKey
        return self
    }
    
    func withCache(timeout: Int?) -> HttpClient {
        self.cacheTimeout = timeout
        return self
    }
    
    func withHeaders(_ headers: Dictionary<String, String>) -> HttpClient {
        self.headers = headers
        return self
    }
    
    func addToQueue<T: Encodable>(uri: String, _ requestBody: T, _ onSuccess: @escaping (Data?) -> Void, _ onError: @escaping (Error) -> Void) {
        if(requestQueue.items.count >= requestQueueLimit) {
            requestQueue.items.forEach {
                makeRequest(uri: $0.uri, $0.body, $0.successCallback, $0.errorCallback)
            }
        }
        
        requestQueue.enqueue(element: DashXRequest(uri: uri, body: requestBody, successCallback: onSuccess, errorCallback: onError))
    }
    
    func makeRequest<T: Encodable>(
        uri: String,
        _ requestBody: T,
        _ onSuccess: @escaping (Data?) -> Void,
        _ onError: @escaping (Error) -> Void
    ) {
        let headerDictionary = headers.merging(
            [ "X-Public-Key": publicKey ?? "" ]
        ) { (current, _) in current }
        
        let headers = HTTPHeaders.init(headerDictionary)
        
        let jsonEncoder = JSONEncoder()
        jsonEncoder.keyEncodingStrategy = .convertToSnakeCase
        
        // For debugging
        if let jsonData = try? jsonEncoder.encode(requestBody),
            let jsonString = String(data: jsonData, encoding: .utf8) {
            DashXLog.d(tag: #function, jsonString)
        }
        
        let dashXRequestInterceptor = DashXRequestInterceptor()
        
        AF.request(
            baseUri! + uri,
            method: .post,
            parameters: requestBody,
            encoder: JSONParameterEncoder(encoder: jsonEncoder),
            headers: headers,
            interceptor: dashXRequestInterceptor
        ) { request in
            if let timeout = self.cacheTimeout {
                request.addValue("private, must-revalidate, max-age=\(String(describing: timeout))", forHTTPHeaderField: "Cache-Control")
            }
        }
        .validate()
        .responseJSON { response in switch response.result {
        case .success:
            onSuccess(response.data)
        case let .failure(error):
            onError(error)
            }
        }
    }
}

File: packages/node/src/HttpClient.ts
import http from 'got'
import Keyv from 'keyv'
import os from 'os'
import type { Response } from 'got'

import { snakeCaseKeys } from './utils'

class HttpClient {
  private getCache: (cacheTimeout:number) => Keyv

  private cacheTimeout?: number

  private publicKey?: string

  private privateKey?: string

  private baseUri: string

  private requestQueue: { uri: string, body: any }[]

  private requestQueueLimit: number

  constructor(baseUri: string, publicKey?: string, privateKey?: string) {
    this.getCache = (cacheTimeout: number) => new Keyv(
      `sqlite://${os.tmpdir}/dashx-response-cache`, { ttl: cacheTimeout }
    )
    this.baseUri = baseUri
    this.publicKey = publicKey
    this.privateKey = privateKey
    this.requestQueue = []
    this.requestQueueLimit = 5
  }

  create(): HttpClient {
    return new HttpClient(this.baseUri, this.publicKey, this.privateKey)
  }

  setCacheTimeout(to?: number): HttpClient {
    this.cacheTimeout = to
    return this
  }

  flushRequestQueue(): Promise<Response>[] {
    return this.requestQueue.map((request) => this.makeRequest(request.uri, request.body))
  }

  addToQueue(uri: string, body: any): Promise<any> {
    if (this.requestQueue.length > this.requestQueueLimit) {
      this.flushRequestQueue()
    }

    this.requestQueue.push({ uri, body })
    return Promise.resolve(body)
  }

  makeRequest(uri: string, body: any): Promise<Response> {
    return http(uri, {
      json: snakeCaseKeys(body),
      method: 'POST',
      prefixUrl: this.baseUri,
      cache: this.cacheTimeout ? this.getCache(this.cacheTimeout) : false,
      headers: {
        'User-Agent': 'dashx-node',
        'X-Public-Key': this.publicKey,
        'X-Private-Key': this.privateKey,
        'Content-Type': 'application/json'
      },
      responseType: 'json'
    })
  }
}

export default HttpClient
